File: Merge Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
bool cmp(const Interval &a, const Interval &b) {
        return a.start < b.start;
    }
class Solution {
public:
    
    vector<Interval> merge(vector<Interval> &intervals) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (intervals.size() < 2) return intervals;
        
        vector<Interval> result;
        sort(intervals.begin(), intervals.end(), cmp);
        Interval cur = intervals[0];
        for (int i = 1; i < intervals.size(); i++) {
            Interval next = intervals[i];
            if (next.end > cur.end) {
                if (next.start <= cur.end) {
                    cur.end = next.end;
                } else {
                    result.push_back(cur);
                    cur = next;
                }
            }
        }
        result.push_back(cur);
        return result;
    }
};

File: Merge Intervals.cs
/**
 * Definition for an interval.
 * public class Interval {
 *     public int start;
 *     public int end;
 *     public Interval() { start = 0; end = 0; }
 *     public Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public IList<Interval> Merge(IList<Interval> intervals) {
        if (intervals == null || intervals.Count <= 1) {
            return intervals;
        }
        
        var result = new List<Interval>();
        intervals = intervals.OrderBy(i => i.start).ToList();
        var cur = intervals[0];
        for (var i = 1; i < intervals.Count; i++) {
            var next = intervals[i];
            if (next.end > cur.end) {
                if (next.start <= cur.end) {
                    cur.end = next.end;
                } else {
                    result.Add(cur);
                    cur = next;
                }
            }
        }
        result.Add(cur);
        
        return result;
    }
}
